### PR TITLE
implement driver metric collector

### PR DESCRIPTION
**Pull request by ImpSy**

### CODE REVIEW

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala [hunk 1]
```diff
@@ -6,6 +6,15 @@ import scala.concurrent.duration._
 
 object Configs {
 
+  private var dmAppId: Option[String] = None
+
+  def getDMAppId(sparkConf: SparkConf): String = {
+    if (dmAppId.isEmpty) {
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: Replace with `Option.getOrElse`:

```
dmAppId.getOrElse(generateDMAppId(sparkConf))
```

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricCollector.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  private def publishMetrics(payload: MetricsPayload): Unit = {
+    val url = s"$collectorURL/metrics"
+
+    try {
+      sendRequest(
+        httpClientMetrics,
+        url,
+        accessTokenOption.get,
+        payload.toJson,
+        "Successfully sent metrics payload"
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricCollector.scala]: nit: Metrics payload sent successfully

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricCollector.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    threads.foreach(_.stop())
+  }
+}
+
+object MetricCollector {
+
+  private var sharedCollector: Option[MetricCollector] = None
+
+  def getOrCreate(hostname: String, sparkConf: SparkConf): MetricCollector = {
+    if (sharedCollector.isEmpty) {
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricCollector.scala]: Use `Option.getOrElse`:

```
sharedCollector.getOrElse { new MetricCollector(hostname, sparkConf) }
```

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsEvent.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      ~ ("timestamp" -> timestamp)
+      ~ ("metrics" -> metrics.toJson))
+  }
+
+  override def toString: String = compact(this.toJson)
+}
+
+object MetricsEvent {
+  def apply(host: String, kind: String, metrics: Metrics): MetricsEvent = {
+    MetricsEvent(host, kind, metrics, currentTime)
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsEvent.scala:26]: Why not making `currentTime` the default value of `timestamp`?

File: common/src/main/scala/co/datamechanics/delight/common/metrics/memory/MemoryMetricsEvent.scala [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package co.datamechanics.delight.common.metrics.memory
+
+import co.datamechanics.delight.common.metrics.MetricsEvent
+
+object MemoryMetricsEvent {
+  def apply(host: String, MemoryMetrics: MemoryMetrics): MetricsEvent = {
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/memory/MemoryMetricsEvent.scala]: `MemoryMetrics` value should be lowercase, no?

**Review by jaceklaskowski — CHANGES_REQUESTED** [review]

**Review by jaceklaskowski — APPROVED** [review]
LGTM

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala [hunk 6]
```diff
@@ -6,6 +6,15 @@ import scala.concurrent.duration._
 
 object Configs {
 
+  private var dmAppId: Option[String] = None
```
Comment (jrj-d) [inline common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: What about:
```
lazy val dmAppId: String = generateDMAppId()
```

Comment (jrj-d) [reply common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: Then  it can be invoked as `Configs.dmAppId`

File: common/src/main/scala/org/apache/spark/PackageProxy.scala [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package org.apache.spark
+
+import org.apache.spark.util.Utils
+
+object PackageProxy {
```
Comment (jrj-d) [inline common/src/main/scala/org/apache/spark/PackageProxy.scala:5]: Can you explain this trick? :)

File: justfile [hunk 8]
```diff
@@ -78,6 +78,26 @@ run_test_app_local_jar spark_distribution_folder spark_version scala_version:
     {{spark_distribution_folder}}/examples/jars/spark-examples_{{scala_version}}-{{spark_version}}.jar \
     100
 
+run_test_app_docker image spark_version scala_version:
```
Comment (jrj-d) [inline justfile:81]: Why did you need this new just target? Compared to run_test_app_local_jar

**Review by jrj-d — APPROVED** [review]
Looks good to me! Can't wait to see this in action

File: common/src/main/scala/org/apache/spark/PackageProxy.scala (hunk 7 above)
Comment (ImpSy) [reply common/src/main/scala/org/apache/spark/PackageProxy.scala:5]: With that trick we can expose private spark packages by defining a new public class PackageProxy

That way we can access all the private packages from the same object as it's define in `common`

File: justfile (hunk 8 above)
Comment (ImpSy) [reply justfile:81]: Because `/proc` is not implemented on OSX thus we need docker to be able to run integration test on the feature

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala (hunk 6 above)
Comment (ImpSy) [reply common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: generateDMAppId() needs access to `sparkConf`, one solution for this would be to change `Configs` to a class that has SparkConf as parameters

File: agent/src/main/scala/co/datamechanics/delight/DelightListener.scala [hunk 9]
```diff
@@ -26,6 +27,9 @@ class DelightListener(sparkConf: SparkConf) extends SparkListener with Logging {
 
   private val streamingConnector = DelightStreamingConnector.getOrCreate(sparkConf)
 
+  private val metricsCollector = MetricCollector.getOrCreate("driver", sparkConf)
```
Comment (mathwall) [inline agent/src/main/scala/co/datamechanics/delight/DelightListener.scala]: Maybe rename the `MetricCollector` class and file into `MetricsCollector`? or rename `metricsCollector` into `metricCollector`

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala [hunk 10]
```diff
@@ -6,6 +6,22 @@ import scala.concurrent.duration._
 
 object Configs {
 
+  def isEdge(sparkConf: SparkConf): Boolean = {
+    sparkConf.getBoolean("spark.delight.edge", false)
+  }
+
+  def getDMAppId(sparkConf: SparkConf): String = {
+    val configPath = "spark.delight.dmAppId"
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: Why is this val called `configPath`? Looks confusing.

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala [hunk 11]
```diff
@@ -6,6 +6,22 @@ import scala.concurrent.duration._
...
+    val dmAppId = sparkConf
+      .getOption(configPath)
+      .getOrElse {
+        val generatedDMAppId = generateDMAppId(sparkConf)
+        sparkConf.set(configPath, generatedDMAppId)
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: This side-effect worries me a lot. Someone outside should take care of doing `set` or it should be a required config or `getDMAppId` should really have a different name.

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala [hunk 12]
```diff
@@ -6,6 +6,22 @@ import scala.concurrent.duration._
 
 object Configs {
 
+  def isEdge(sparkConf: SparkConf): Boolean = {
+    sparkConf.getBoolean("spark.delight.edge", false)
+  }
+
+  def getDMAppId(sparkConf: SparkConf): String = {
+    val configPath = "spark.delight.dmAppId"
+    val dmAppId = sparkConf
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: Remove `dmAppId` and leave the expression alone.

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import scala.concurrent.duration.{FiniteDuration, SECONDS}
+
+class MetricsCollector(hostname: String, sparkConf: SparkConf) extends Logging {
+  private val dmAppId = DmAppId(Configs.getDMAppId(sparkConf))
+  private val collectorURL = Configs.collectorUrl(sparkConf).stripSuffix("/")
+  private val accessTokenOption = Configs.accessTokenOption(sparkConf)
+
+  private val started: AtomicBoolean = new AtomicBoolean(false)
+  private val MetricsQueue = new mutable.Queue[MetricsEvent]()
+  private val httpClientMetrics: HttpClient = new DefaultHttpClient()
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala]: nit: Is this type declaration required here?

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+      threads += startRepeatThread(FiniteDuration(2, SECONDS)) {
+        logDebug("Logged metrics Sender")
+        sendMetrics()
+      }
+      logInfo("Started MemoryMetrics Sender thread")
+    }
+  }
+
+  def stop(): Unit = {
+    threads.foreach(_.stop())
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala:91]: Should these threads be `null`ed too to avoid a memory leak? Just checking...

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+  }
+}
+
+object MetricsCollector {
+
+  private var sharedCollector: Option[MetricsCollector] = None
+
+  def getOrCreate(hostname: String, sparkConf: SparkConf): MetricsCollector = {
+    if (sharedCollector.isEmpty) {
+      sharedCollector = Option(new MetricsCollector(hostname, sparkConf))
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala]: Wonder if making `sharedCollector` a `lazy val` wouldn't be the same yet less code?

```
lazy val sharedCollector = new MetricsCollector(hostname, sparkConf)
```

File: common/src/main/scala/co/datamechanics/delight/common/package.scala [hunk 16]
```diff
@@ -2,4 +2,4 @@ package co.datamechanics.delight
 
 import org.apache.spark.internal.Logging
 
-package object common extends Logging {}
\ No newline at end of file
+package object common extends Logging {}
```
Comment (jaceklaskowski) [inline common/src/main/scala/co/datamechanics/delight/common/package.scala]: What's the purpose of `common` to extend `Logging`? Can't think of a use case for this (and can't wait to learn a new trick in Scala).

**Review by jaceklaskowski — CHANGES_REQUESTED** [review]

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala (hunk 11 above)
Comment (ImpSy) [reply common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: This will be used soon when we release our executor plugin
We'll need to share a same `dmAppId` across numerous process and the `sparkConf` is the easiest way to get everything sync together
The 1st element needing the `dmAppId` will generate it and set it in the config every other process will now pick it up from there

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala (hunk 15 above)
Comment (ImpSy) [reply common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala]: How do you pass the params to the object  if you want to use a lazy val ?

File: common/src/main/scala/co/datamechanics/delight/common/package.scala (hunk 16 above)
Comment (ImpSy) [reply common/src/main/scala/co/datamechanics/delight/common/package.scala]: This `extends` is a bit of legacy that stays after the package split
Thanks to have noticed it, i'll remove it

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala (hunk 14 above)
Comment (ImpSy) [reply common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala:91]: This method is called only when the app terminates, i don't think that we have to monitor this as the JVM profil will also stop

File: common/src/main/scala/co/datamechanics/delight/common/Configs.scala (hunk 11 above)
Comment (jaceklaskowski) [reply common/src/main/scala/co/datamechanics/delight/common/Configs.scala]: Mind to create an issue / task to track it?

File: common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala (hunk 15 above)
Comment (jaceklaskowski) [reply common/src/main/scala/co/datamechanics/delight/common/metrics/MetricsCollector.scala]: Correct! Can you change `Option` to `Some`?

**Review by jaceklaskowski — APPROVED** [review]
LGTM
